modules/gdscript/editor/script_templates/CharacterBody3D/basic_movement.gd

```diff
@@ -4,27 +4,38 @@ extends _BASE_
 
 
 const SPEED = 5.0
+const ACCEL = 5.0
 const JUMP_VELOCITY = 4.5
 
 
 func _physics_process(delta: float) -> void:
-	# Add the gravity.
+	# Handle gravity.
 	if not is_on_floor():
 		velocity += get_gravity() * delta
 
+	# Get the vertical velocity (for jump).
+	var vertical_velocity: Vector3 = velocity.project(up_direction)
+
+	# Get the horizontal velocity (for acceleration).
+	var horizontal_velocity: Vector3 = velocity - vertical_velocity
+
 	# Handle jump.
 	if Input.is_action_just_pressed("ui_accept") and is_on_floor():
-		velocity.y = JUMP_VELOCITY
-
-	# Get the input direction and handle the movement/deceleration.
-	# As good practice, you should replace UI actions with custom gameplay actions.
-	var input_dir := Input.get_vector("ui_left", "ui_right", "ui_up", "ui_down")
-	var direction := (transform.basis * Vector3(input_dir.x, 0, input_dir.y)).normalized()
-	if direction:
-		velocity.x = direction.x * SPEED
-		velocity.z = direction.z * SPEED
-	else:
-		velocity.x = move_toward(velocity.x, 0, SPEED)
-		velocity.z = move_toward(velocity.z, 0, SPEED)
+		vertical_velocity = up_direction * JUMP_VELOCITY
+
+	# Get the 2D input vector. As good practice, you should replace UI actions with custom gameplay actions.
```

Comment:
I would normally not say this but given the context it's probably worth to split these in two.
```suggestion
	# Get the 2D input vector.
	# As good practice, you should replace UI actions with custom gameplay actions.
```
Which has one realise, the comment is repeating what the code already says fairly explicitly. So there's no point.
```suggestion
	# As good practice, you should replace UI actions with custom gameplay actions.
```

Reply:
Thank you for the review! I'll work on those!

About the comments, i added them because i wanted to be a bit more explicit for new users, as i often see them having issues to understand some of this things when they start. However, you are totally right that the code is pretty self explanatory (I tried to separate the steps and use clear variable names for that end). I don't feel particularly strong towards keeping them or removing them, but i do think they can be useful. Maybe we can simplify them so they are less redundant?

In any case, if you think the code is self explanatory enough, i have no issue with fully removing them!

Waiting for your feedback.

Thanks a lot!

Reply:
I have no suggestions for the moment, but generally speaking code comments are _really_ good when they explain why the code exists, rather than what it does. Unorganized thoughts ahead:

- When getting both the horizontal and vertical velocity, it may be worth explaining why separating the two in the first place. It is unorthodox, and it doesn't help that neither values are `float`, rather they're `Vector3`.
    - Instead of `# Get the vertical velocity (for jump).`, there could be a few details on what project does, instead. May be too much, though.

- There's always exceptions, of course. I think that `# Calculate the intended direction in 3D space.` is a good comment there because the it answers **why** we need the line below, instead of exactly **what** it does step-by-step, if you think about it. The math is also particularly unfriendly, even more so for beginners.

Reply:
Those are good thoughts! I just pushed a less verbose version of the comments, including your previous recommendations, it looks more clean now. I'm going to revisit the comments and try to see if i can explain exactly why we do each step without adding too much text.

I don't think the math is unfriendly in the context of what it does. I agree that has an inherent complexity but it is the easiest way i could think of to illustrate how to perform correct acceleration/friction on the horizontal axis without it being affected or disturbing the vertical one.

I'm gonna think a bit about it and see if i can make the math more intuitive by explaining the reasoning in the comments instead of being so literal. I think that will help a lot.

Thank you a lot for the help! really appreciate it!

Reply:
Regarding the math, I didn't mean it should be changed. I wanted to say that summarizing it with a comment is a good thing.

Reply:
I have this version now, it's a bit more verbose but it may be more clear to read. Before committing, who do you see it against the current one? The issue i have is basically that, to explain properly what is going on it requires a lot more text and context, and that's bad for readability. 

I like this version, but having the cleaner one and a proper documentation page for the templates may be a good thing. I could work on that if needed.

Thank you!

```
func _physics_process(delta: float) -> void:
	# Handle gravity.
	if not is_on_floor():
		velocity += get_gravity() * delta

	# Isoltate the vertical and horizontal components of the velocity so we can prevent
	# any vertical velocity from affecting our horizontal acceleration/friction.
	var vertical_velocity: Vector3 = velocity.project(up_direction)
	var horizontal_velocity: Vector3 = velocity - vertical_velocity

	# Handle jump.
	if Input.is_action_just_pressed("ui_accept") and is_on_floor():
		vertical_velocity = up_direction * JUMP_VELOCITY

	# As good practice, you should replace UI actions with custom gameplay actions.
	var input_vector := Input.get_vector("ui_left", "ui_right", "ui_up", "ui_down", 0.15)

	# Calculate the intended direction in 3D space.
	var input_direction: Vector3 = transform.basis.orthonormalized() * Vector3(input_vector.x, 0, input_vector.y)

	# input_direction is usually a unit vector, but it's magnitude can be less than 1 to allow the
	# use of analog input devices. As such, no special code is needed to handle controller joystic input.
	var target_horizontal_velocity: Vector3 = input_direction * SPEED

	# Move the current horizontal velocity towards the target horizontal velocity.
	horizontal_velocity = horizontal_velocity.move_toward(target_horizontal_velocity, ACCEL * delta)

	# Compose the final velocity.
	velocity = horizontal_velocity + vertical_velocity
```